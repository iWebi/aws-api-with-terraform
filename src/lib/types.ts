export interface BaseEntity {
  hashKey: string;
  rangeKey?: string;
  created: number;
  updated: number;
}

export interface Account extends BaseEntity {
  tenantId: string;
  id: string;
  name: string;
}

export interface AppError {
  body: string;
  statusCode: number;
  statusType: string;
}

export interface DynamoItemResponse<T> {
  body: T;
  statusCode: number;
  statusType: string;
}
