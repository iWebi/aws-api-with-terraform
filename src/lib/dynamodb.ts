import { DeleteItemOutput, DynamoDBClient, UpdateItemOutput } from "@aws-sdk/client-dynamodb";
import { DeleteCommand, DynamoDBDocumentClient, GetCommand, PutCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import * as AWSXRay from "aws-xray-sdk-core";
import { DynamoItemResponse } from "./types";
import {debug, internalServerErrorWith, notFoundResponse, okResponse, setStatusType} from "./utils";
import {PutCommandInput} from "@aws-sdk/lib-dynamodb/dist-types/commands/PutCommand";
import {UpdateCommandInput} from "@aws-sdk/lib-dynamodb/dist-types/commands/UpdateCommand";
import {GetCommandInput} from "@aws-sdk/lib-dynamodb/dist-types/commands/GetCommand";

const ddbDocClient = buildDocClient();

export async function getItem<T>(params: GetCommandInput): Promise<DynamoItemResponse<T>> {
  const data = await ddbDocClient.send(new GetCommand(params));
  if (!data.Item) {
    throw404(params);
  }
  return okResponse(data.Item! as T);
}

export async function putItem<T>(params: PutCommandInput): Promise<DynamoItemResponse<T>> {
  try {
    await ddbDocClient.send(new PutCommand(params));
  } catch (err) {
    throw errorResponse(err);
  }
  return okResponse(params.Item as T);
}

export async function updateItem(params: UpdateCommandInput): Promise<DynamoItemResponse<any>> {
  let data: UpdateItemOutput = {};
  try {
    data = await ddbDocClient.send(new UpdateCommand(params));
  } catch (err) {
    throw errorResponse(err);
  }
  return okResponse([data.Attributes!]);
}

export async function deleteItem(params: any): Promise<DynamoItemResponse<any>> {
  let data: DeleteItemOutput = {};
  try {
    data = await ddbDocClient.send(new DeleteCommand(params));
  } catch (err) {
    throw errorResponse(err);
  }
  if (!data.Attributes) {
    throw404(params);
  }
  return okResponse(data.Attributes!);
}

function errorResponse(err: any) {
  if (err.hasOwnProperty("$metadata")) {
    const code = err.httpStatusCode;
    return {
      body: err.message,
      statusCode: code,
      statusType: setStatusType(code),
    };
  }
  return internalServerErrorWith(JSON.stringify(err));
}

function throw404(params: any, message?: string) {
  debug("Item not found for params", params);
  throw notFoundResponse(message);
}

export function buildDocClient() {
  const marshallOptions = {
    // Whether to automatically convert empty strings, blobs, and sets to `null`.
    convertEmptyValues: false, // false, by default.
    // Whether to remove undefined values while marshalling.
    removeUndefinedValues: true, // false, by default.
    // Whether to convert typeof object to map attribute.
    convertClassInstanceToMap: false, // false, by default.
  };

  const unmarshallOptions = {
    // Whether to return numbers as a string instead of converting them to native JavaScript numbers.
    wrapNumbers: false, // false, by default.
  };

  const { IS_OFFLINE } = process.env;
  const offlineOptions = {
    region: "localhost",
    endpoint: "http://localhost:8000",
    // either .aws/credentials or ENV variables or config needs credentials keys.
    // These are checked by AWS SDK but not used in offline mode
    credentials: {
      accessKeyId: "required-but-not-used-weird",
      secretAccessKey: "required-but-not-used-weird",
    },
  };
  const dynamodb = IS_OFFLINE
    ? new DynamoDBClient(offlineOptions)
    : AWSXRay.captureAWSv3Client(
        new DynamoDBClient({
          region: process.env.AWS_REGION,
        })
      );
  return DynamoDBDocumentClient.from(dynamodb, { marshallOptions, unmarshallOptions });
}
