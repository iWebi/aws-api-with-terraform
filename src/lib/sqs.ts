import { SendMessageCommand, SendMessageCommandInput, SQSClient } from "@aws-sdk/client-sqs";
import * as AWSXRay from "aws-xray-sdk-core";
import {Account} from "./types";

const sqsClient = buildSQSClient();

function buildSQSClient() {
  const { IS_OFFLINE } = process.env;
  const offlineOptions = {
    region: "localhost",
    endpoint: "http://localhost:9324",
  };
  return IS_OFFLINE
    ? new SQSClient(offlineOptions as any)
    : AWSXRay.captureAWSv3Client(new SQSClient({} as any));
}

function getNotificationsQueueName(): string {
  const { IS_OFFLINE, NOTIFICATION_QUEUE_URL } = process.env;
  return IS_OFFLINE ? "http://localhost:9324/queue/accountnotifications.fifo" : NOTIFICATION_QUEUE_URL!;
}

export async function sendNewAccountMessage(account: Account) {
  const sendMessageInput = {
    QueueUrl: getNotificationsQueueName()!,
    MessageBody: JSON.stringify(account),
    MessageGroupId: `Account-${account.id}-${account.tenantId}`,
  } as SendMessageCommandInput;
  await sqsClient.send(new SendMessageCommand(sendMessageInput));
}
