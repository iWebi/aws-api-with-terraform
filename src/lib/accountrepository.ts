import { DeleteCommandInput, GetCommandInput, PutCommandInput } from "@aws-sdk/lib-dynamodb";
import { ulid } from "ulid";
import { ACCOUNT_TABLE } from "./constants";
import { deleteItem, getItem, putItem } from "./dynamodb";
import { Account, DynamoItemResponse } from "./types";
import { okResponse } from "./utils";

export async function addAccount(account: Account): Promise<DynamoItemResponse<Account>> {
  account.id = ulid();
  account.hashKey = hashKeyFor(account);
  account.rangeKey = account.id;
  account.created = account.updated = new Date().getTime();
  const params = {
    Item: account,
    TableName: ACCOUNT_TABLE,
  } as PutCommandInput;

  await putItem(params);
  return okResponse(account);
}

function hashKeyFor(account: Account): string {
  return `Account-${account.tenantId}`;
}

export async function getAccount(accountId: string, tenant: string): Promise<DynamoItemResponse<Account>> {
  const params = {
    Key: {
      hashKey: `Account-${tenant}`,
      rangeKey: accountId,
    },
    TableName: ACCOUNT_TABLE,
  } as GetCommandInput;
  return await getItem(params);
}

export async function updateAccount(account: Account): Promise<DynamoItemResponse<Account>> {
  account.hashKey = hashKeyFor(account);
  account.rangeKey = account.id;
  account.updated = new Date().getTime();

  const params = {
    Item: account,
    TableName: ACCOUNT_TABLE,
  } as PutCommandInput;

  await putItem(params);
  return okResponse(params.Item! as Account);
}

export async function deleteAccount(accountId: string, tenant: string): Promise<DynamoItemResponse<Account>> {
  const params = {
    Key: {
      hashKey: `Account-${tenant}`,
      rangeKey: accountId,
    },
    TableName: ACCOUNT_TABLE,
    ReturnValues: "ALL_OLD",
  } as DeleteCommandInput;
  return await deleteItem(params);
}
