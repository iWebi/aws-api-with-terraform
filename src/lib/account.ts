import { APIGatewayProxyResult } from "aws-lambda";
import * as repository from "./accountrepository";
import {Account, AppError} from "./types";
import { badRequestWith, errorResponse, isEmpty, successResponse } from "./utils";

export async function addAccount(accountRequest: string, tenant: string): Promise<APIGatewayProxyResult> {
  try {
    const account = validateAccountForCreateOrUpdate(accountRequest);
    account.tenantId = tenant;
    await repository.addAccount(account);
    // await sendNewAccountMessage(account);
    return successResponse(account, 200);
  } catch (err) {
    return errorResponse(err as AppError);
  }
}

function validateAccountForCreateOrUpdate(accountRequest: string): Account {
  let account: Account;
  try {
    account = JSON.parse(accountRequest);
  } catch (err) {
    throw badRequestWith("Invalid JSON payload");
  }
  // Dummy validations. JSON schema validators https://ajv.js.org can be considered instead
  if (isEmpty(account.name)) {
    throw badRequestWith("Missing mandatory field(s): name for account creation");
  }
  return account;
}

export async function getAccount(accountId: string, tenant: string): Promise<APIGatewayProxyResult> {
  try {
    const accountResponse = await repository.getAccount(accountId, tenant);
    return successResponse(accountResponse.body, 200);
  } catch (err) {
    return errorResponse(err as AppError);
  }
}

export async function deleteAccount(accountId: string, tenant: string): Promise<APIGatewayProxyResult> {
  try {
    const deleteResponse = await repository.deleteAccount(accountId, tenant);
    return successResponse(deleteResponse.body, 200);
  } catch (err) {
    return errorResponse(err as AppError);
  }
}

export async function updateAccount(
  accountRequest: string,
  accountId: string,
  tenant: string
): Promise<APIGatewayProxyResult> {
  try {
    const account = validateAccountForCreateOrUpdate(accountRequest);
    account.id = accountId;
    account.tenantId = tenant;
    const updateResponse = await repository.updateAccount(account);
    return successResponse(updateResponse.body);
  } catch (err) {
    return errorResponse(err as AppError);
  }
}
