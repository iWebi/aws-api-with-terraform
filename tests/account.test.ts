import * as ulid from "ulid";
import {addAccount} from "../src/lib/account";
import {AccountRequest} from "./types";
jest.mock("../src/lib/dynamodb");
jest.mock("ulid");

const mockedUlid = ulid as jest.Mocked<typeof ulid>;
const randomness = "test-random-01HB4Y3WPZ965E7D6G8JG9610A";
mockedUlid.ulid.mockReturnValue(randomness);

describe("addAccount", () => {
    test("fails for invalid JSON request data", async () => {
        // when
        const response = await addAccount("foo", "12345");
        // then
        expect(response.statusCode).toBe(400);
        expect(response).toMatchSnapshot();
    });

    test("fails for missing mandatory fields in request data", async () => {
        // when
        const response = await addAccount('{"foo": "bar"}', "12345");
        // then
        expect(response.statusCode).toBe(400);
        expect(response).toMatchSnapshot();
    });

    test("creates account for valid request data", async () => {
        // given
        const tenantId = "tenant:12345";
        const accountReq:AccountRequest = {
            name: "test-account-name-"+new Date().getTime()
        }
        // when
        const response = await addAccount(JSON.stringify(accountReq), tenantId);
        // then
        expect(response.statusCode).toBe(200);
        const accountCreated = JSON.parse(response.body);
        expect(accountCreated).toMatchObject({
            id: randomness,
            name: accountReq.name,
            tenantId
        });
    });
});
