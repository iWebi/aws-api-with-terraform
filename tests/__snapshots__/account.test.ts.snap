// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addAccount fails for invalid JSON request data 1`] = `
{
  "body": "{"body":"Invalid JSON payload","statusCode":400,"statusType":"Bad Request"}",
  "headers": {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`addAccount fails for missing mandatory fields in request data 1`] = `
{
  "body": "{"body":"Missing mandatory field(s): name for account creation","statusCode":400,"statusType":"Bad Request"}",
  "headers": {
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;
